core/arch/arm/plat-imx/drivers/imx_csu.c

```diff
@@ -1,135 +1,239 @@
+	{ 39, 0x3300FF }, /* Protect OCRAM */
+	{ (-1), 0 },
 };
 
-const struct csu_setting csu_setting_imx6ull[] = {
-	{ 13, 0xFF0033 },	/* Protect ROMCP */
-	{ 16, 0x3300FF },	/* Protect TZASC */
-	{ 34, 0xFF0033 },	/* Protect DCP */
-	{ 39, 0x3300FF },	/* Protect OCRAM */
+const struct csl_cfg csl_imx6sl[] = {
+	{ 13, 0x3F0033 }, /* Protect DCP/ROMCP */
+	{ 16, 0xFF0033 }, /* Protect TZASC */
+	{ 26, 0xFF0033 }, /* Protect OCRAM */
 	{ (-1), 0 },
 };
 
-const struct csu_setting csu_setting_imx6sl[] = {
-	{ 13, 0x3F0033 },	/* Protect DCP/ROMCP */
-	{ 16, 0xFF0033 },	/* Protect TZASC */
-	{ 26, 0xFF0033 },	/* Protect OCRAM */
+const struct csl_cfg csl_imx6sx[] = {
+	{ 13, 0xFF0033 }, /* Protect ROMCP */
+	{ 15, 0xFF0033 }, /* Protect RDC */
+	{ 16, 0x3300FF }, /* Protect TZASC */
+	{ 34, 0x3300FF }, /* Protect OCRAM */
 	{ (-1), 0 },
 };
 
-const struct csu_setting csu_setting_imx6sx[] = {
-	{13, 0xFF0033},		/* Protect ROMCP */
-	{15, 0xFF0033},		/* Protect RDC   */
-	{16, 0x3300FF},		/* Protect TZASC */
-	{34, 0x3300FF},		/* Protect OCRAM */
-	{(-1), 0},
+const struct csl_cfg csl_imx7ds[] = {
+	{ 14, 0x3300FF }, /* Protect RDC */
+	{ 15, 0xFF0033 }, /* Protect CSU */
+	{ 28, 0xFF0033 }, /* Protect TZASC */
+	{ 59, 0x3300FF }, /* Protect OCRAM_S */
+	{ (-1), 0 },
 };
 
-const struct csu_setting csu_setting_imx7ds[] = {
-	{14, 0x3300FF},		/* Protect RDC     */
-	{15, 0xFF0033},		/* Protect CSU     */
-	{28, 0xFF0033},		/* Protect TZASC   */
-	{59, 0x3300FF},		/* Protect OCRAM_S */
-	{(-1), 0},
+struct csl_cfg_node {
+	void (*configure)(vaddr_t va);
+	SLIST_ENTRY(csl_cfg_node) next;
 };
 
-/* Set all masters to non-secure except the Cortex-A7 */
-const struct csu_sa_setting csu_sa_imx6ul = { 0x10554550, 0x20aa8aa2 };
+SLIST_HEAD(csl_runtime_cfg, csl_cfg_node);
+struct csu_config {
+	const struct csl_cfg *csl;
+	const struct sa_cfg *sa;
+	struct csl_runtime_cfg csl_list;
+	vaddr_t va;
+};
 
-const struct csu_config csu_imx6 = { NULL, csu_setting_imx6 };
-const struct csu_config csu_imx6ul = { &csu_sa_imx6ul, csu_setting_imx6ul };
-const struct csu_config csu_imx6ull = { NULL, csu_setting_imx6ull };
-const struct csu_config csu_imx6sl = { NULL, csu_setting_imx6sl };
-const struct csu_config csu_imx6sx = { NULL, csu_setting_imx6sx };
-const struct csu_config csu_imx7ds = { NULL, csu_setting_imx7ds };
+/* Default Configuration:
+ *
+ *  Defines all the different default configuration options for the different
+ *  systems-on-chip.
+ */
+static const struct default_cfg {
+	bool (*found)(void);
+	const struct csu_config cfg;
+} default_cfg[] = {
+	{ soc_is_imx6sx, { csl_imx6sx, NULL }, },
+	{ soc_is_imx6ul, { csl_imx6ul, &sa_imx6ul }, },
+	{ soc_is_imx6ull, { csl_imx6ull, NULL }, },
+	{ soc_is_imx6sll, { csl_imx6sl, NULL }, },
+	{ soc_is_imx6sl, { csl_imx6sl, NULL }, },
+	{ soc_is_imx6, { csl_imx6, NULL }, },
+	{ soc_is_imx7ds, { csl_imx7ds, NULL }, },
+};
 
-static TEE_Result csu_init(void)
+static bool rngb_is_enabled(void)
```

Comment:
I did not expect a rework of the CSU in this pull-request. When I suggested changes in
https://github.com/OP-TEE/optee_os/pull/4415#discussion_r586401954
I was expecting a feedback on my suggestion, not a rework.

The changes I suggested aims at removing any kind of dependency to other drivers like this function, and instead, have a function like ```imx_csu_add()``` you could call from another driver so we don't have to modify ```imx_csu.c``` every time we want to protect something.
I don't really see what's new with this rework compared to what we had before.

Maybe we should dedicate another PR for this?
So you can get the RNG driver merged quickly?

Reply:
ok, I didnt think this was actually maintained (btw I did ask who was maintaining imx_csu but got no answer hence the rework - I guess tracking questions on github is as hard for everyone else as it is for me). So the main flaw of the  old driver to support these new requirements was that it was not open for extensions. The new driver addresses this by using a linked list and an array where to plug new functionality. This can be further enhanced by using different files to link the driver if extending (not modifying) the imx_csu is a concern. But sure, if you dont like it just drop it...it is just code :)

imx_csu_add(): what knowledge will it contain? wont we need to synchronize it with the boot sequence of other drivers.or will this just be posting some generic command to list to execute once the driver has been initialized? maybe better if the maintainer/s take a stab and I just use the new interface.

Reply:
I should have answered you but I was on leave last week. Yes, it might not be obvious but CSU driver is maintained by ```plat-imx``` maintainers.
Ok, I see. Anyway, I wasn't expecting a rework here. You can go back to what you did previously.

My idea was to have
```imx_csu_init(void); early_init?(imx_csu_init)``` : to initialize the csu driver context (get the vaddr_t, check if registers are not already locked ...)
```imx_csu_add(unsigned int index, unint32_t csl_shift, uint16_t permissions)``` : to write to CSL registers from the differents drivers ; depending on whether the driver is enabled or not/successfully initialized or not. There is no need to write all CSL registers at once.
```imx_csu_lock(void); driver_init_late?(imx_csu_lock)``` : to lock the configuration like we do eventually for 6ul.

Reply:
um not sure if this is a good idea - it kind of breaks the driver encapsulation doesnt it? I mean users (ie the RNGB driver) will need to know of the guts of the CSU (different csl shifts for the different supported CSUs).

Reply:
Yes, it breaks the kind of 'encapsulation' of the CSU driver. But since the 'user', the RNG driver here, is supposed to be trusted, I don't know if it's really an issue.

Reply:
IMO is a software design issue but in any case, please let me know how you wish to handle this. the csu discussion seems to be blocking the rngb merge

Reply:
Please revert to what you did initially, I will give you my tag to get this PR merged.
I will submit something for the CSU driver and I will ask for your feedback.

Reply:
thanks @clementfaure